### PR TITLE
Add ability to change column visibility using a proxy

### PR DESCRIPTION
**Pull request by gergness**
I wanted more fine grain control over setting column visibility than is allowed in the ColVis Button add-in from the datatables library. 

This PR adds the functions `showCols` and `hideCols` which work on a proxy object. I'm not sure the API for this is exactly right yet, perhaps there should also be a `toggleColVis` function which allows you to set the visibility to the opposite of what it currently is, so please let me know what you think.

Thank you for your consideration (and all the great work on the package in general)!

### CODE REVIEW

**Comment by gergness** [comment]
I realized I also want the ability to reorder, which I've now included as part of this PR. Is it possible for me as a DT user to pass custom messages to the proxy? I couldn't figure out how, but that might be a better API.

**Comment by shrektan** [comment]
Looks good. However, when testing it, the example in `inst/example/DT-proxy/` didn't work correctly so I was distracted to the fix #528 ... Will take a closer look tomorrow.

**Comment by shrektan** [comment]
@yihui I'm ok with this PR. It's simple and clean, except the seemed redundant `hideColumns()`.

Although it's not necessary, I build a test app for this PR. Didn't find any issues.

```r
library(shiny)
library(DT)

test_cases <- list(
  showAllCols = quote({
    DT::showCols(tbl1, 0:5)
  }),
  showAllCols_reset_works = quote({
    DT::showCols(tbl1, 0, TRUE)
    DT::showCols(tbl2, 0, TRUE)
  }),
  showAllCols_scalar_string = quote({
    DT::showCols(tbl1, "2", TRUE)
    DT::showCols(tbl2, "2", TRUE)
  }),
  showAllCols_string_vector = quote({
    DT::showCols(tbl1, c("2", "4"), TRUE)
    DT::showCols(tbl2, c("2", "4"), TRUE)
  }),
  showAllCols_negtive = quote({
    DT::showCols(tbl1, c(-1:1), TRUE)
    DT::showCols(tbl2, c(-1:1), TRUE)
  }),
  showAllCols_out_of_bounds = quote({
    DT::showCols(tbl1, 5:8, TRUE)
    DT::showCols(tbl2, 5:8, TRUE)
  }),
  showAllCols_out_of_bounds2 = quote({
    DT::showCols(tbl1, 8, TRUE)
    DT::showCols(tbl2, 8, TRUE)
  }),
  hideColumns = quote({
    # hideColumns() doesn't not work
    DT::hideColumns(tbl1, 1:2)
    DT::hideColumns(tbl2, 1:2)
  }),
  hideCols1_scalar = quote({
    DT::hideCols(tbl1, 3, F)
    DT::hideCols(tbl2, 3, F)
  }),
  hideCols2_vector = quote({
    DT::hideCols(tbl1, c(1, 4), F)
    DT::hideCols(tbl2, c(1, 4), F)
  }),
  hideCols_reset_works1 = quote({
    DT::hideCols(tbl1, 1, T)
    DT::hideCols(tbl2, 1, T)
  }),
  hideCols_reset_works2 = quote({
    DT::hideCols(tbl1, c(2, 3), T)
    DT::hideCols(tbl2, c(2, 3), T)
  }),
  hideCols_out_of_bounds = quote({
    DT::hideCols(tbl1, 5:6, T)
    DT::hideCols(tbl2, 5:6, T)
  }),
  hideCols_out_of_bounds2 = quote({
    DT::hideCols(tbl1, 6, T)
    DT::hideCols(tbl2, 6, T)
  }),
  hideCols_string_input = quote({
    DT::hideCols(tbl1, "5", T)
    DT::hideCols(tbl2, "5", T)
  }),
  colReorder_origOrder_TRUE = quote({
    DT::colReorder(tbl1, c(3, 2, 1, 5, 4, 0), TRUE)
    DT::colReorder(tbl2, c(3, 2, 1, 5, 4, 0), TRUE)
  }),
  colReorder_origOrder_FALSE = quote({
    DT::colReorder(tbl1, 5:0, FALSE)
    DT::colReorder(tbl2, 5:0, FALSE)
  }),
  # works but strangely
  colReorder_more_than_ncol = quote({
    DT::colReorder(tbl1, 6:0, TRUE)
    DT::colReorder(tbl2, 6:0, TRUE)
  }),
  # won't work but doesn't matter I think
  colReorder_miss_one_col = quote({
    DT::colReorder(tbl1, 4:0, TRUE)
    DT::colReorder(tbl2, 4:0, TRUE)
  }),
  # won't work but doesn't matter I think
  colReorder_miss_string_input = quote({
    DT::colReorder(tbl1, as.character(5:0), TRUE)
    DT::colReorder(tbl2, as.character(5:0), TRUE)
  })
)

dat <- iris
colnames(dat) <- paste0("COL_", 1:5)

ui = fluidPage(
  inputPanel(
    selectInput("test_id", label = "test id", choices = names(test_cases)),
    uiOutput("test_desc"),
    actionButton("test_run", "run test")
  ),
  fluidRow(
    column(6, h1("server = TRUE"), DTOutput("tbl1")),
    column(6, h1("server = FALSE"), DTOutput("tbl2"))
  )
)

server = function(input, output, session) {
  output$tbl1 = renderDT(
    dat, server = TRUE, extensions = 'ColReorder', options = list(colReorder = TRUE)
  )
  output$tbl2 = renderDT(
    dat, server = FALSE, extensions = 'ColReorder', options = list(colReorder = TRUE)
  )
  tbl1 = DT::dataTableProxy("tbl1")
  tbl2 = DT::dataTableProxy("tbl2")
  output$test_desc <- renderUI({
    case <- test_cases[[input$test_id]]
    txt <- capture.output(print(case))
    do.call(tagList, lapply(txt, tags$p))
  })
  observeEvent(input$test_run, {
    eval(test_cases[[input$test_id]])
  })
}

runApp(mget(c("ui", "server")), port = 13338)
```

File: inst/htmlwidgets/datatables.js [hunk 1]
```diff
@@ -1081,6 +1081,24 @@ HTMLWidgets.widget({
       table.draw();
     }
 
+    methods.hideCols = function(hide, reset) {
+      if (reset) {
+        $(table.columns().visible(true, false))
```
Comment (shrektan) [inline inst/htmlwidgets/datatables.js]: @yihui I think `table.columns().visible(true, false);` should be ok. Am I right?

Comment (yihui) [reply inst/htmlwidgets/datatables.js]: Good catch. You are correct. There is no need to use `$()`.

Comment (shrektan) [reply inst/htmlwidgets/datatables.js]:
```js
methods.showCols = function(show, reset) {
      if (reset) table.columns().visible(false, false);
      table.columns(show).visible(true);
}
```
I think @yihui will like this style but I can't edit the PR directly.

Comment (yihui) [reply inst/htmlwidgets/datatables.js]: Right. Usually I don't use curly braces if there is only a single expression in them, but I'm fine with them.

**Comment by gergness** [comment]
Thanks for reviewing! I found a bug related to filtering after reordering, but it affects columns that have been reordered when the user uses their mouse, so I'll report it separately.

**Pull request merged**